libqtile/backend/x11/core.py

```diff
@@ -559,30 +593,89 @@ def handle_FocusOut(self, event) -> None:
+        sel = self._selection[name]
+
+        if sel.status == SelectionStatus.targets:
+            prop = self._selection_window.get_property(event.selection, xcffib.xproto.Atom.ATOM, int)
+            if prop:
+                if not sel.targets:
+                    sel.targets = [self.conn.atoms.get_name(at) for at in prop]
+                    if "UTF8_STRING" in sel.targets:
+                        idx = sel.targets.index("UTF8_STRING")
+                        sel.targets[0], sel.targets[idx] = sel.targets[idx], sel.targets[0]
+                    sel._cur_tar = 0
+                elif sel._cur_tar < len(sel.targets):
+                    sel._cur_tar += 1
+
+                if sel._cur_tar == len(sel.targets):
+                    logger.debug(f"All reported targets were invalid for {sel}.")
+                    sel.status = SelectionStatus.empty
+                    self.conn.conn.core.DeleteProperty(self._selection_window.wid, event.selection)
+                else:
+                    self.convert_selection(event.selection, sel.current_target)
+            elif sel.targets:
+                logger.debug(f"Settled on target {sel.current_target} for {sel}.")
+                sel.status = SelectionStatus.awaiting
+            else:
+                logger.warn(f"Selection requested targets, but did not receive them. {sel}")
+                sel.status = SelectionStatus.empty
+        if sel.status == SelectionStatus.awaiting:
+            # We need to determine if we need to go into incremental transfer mode.
+            prop = self._selection_window.get_property(event.selection, xcffib.xproto.Atom.Any, dict)
+            if not prop:
+                logger.warn(f"Before copying, selection was unavailable. {sel}")
+                sel.status = SelectionStatus.empty
+            if self.conn.atoms.get_name(prop["type"]) == "INCR":
+                # Initiate the incremental transfer
+                sel.status = SelectionStatus.incremental_transfer
+                self.conn.conn.core.DeleteProperty(self._selection_window.wid, event.selection)
+            else:
+                prop = self._selection_window.get_property(event.selection, prop["type"])
+                sel.binary_selection = b"".join(prop.value)
+                try:
+                    value = sel.binary_selection.decode()
+                except UnicodeDecodeError:
+                    value = ""
+                sel.selection = value
+                sel.status = SelectionStatus.available
+        elif getattr(event, "owner", None) and sel.status in [SelectionStatus.empty, SelectionStatus.available]:
+            sel = self._selection[name] = Selection()
+            sel.status = SelectionStatus.targets
+            self.convert_selection(event.selection, "TARGETS")
+
+        if sel.status == SelectionStatus.available:
+            hook.fire("selection_change", name, sel)
+
+        hook.fire("selection_notify", name, sel)
```

Comment:
I realised we passed equivalent args to the hooks even before this PR, but are they really useful? My concern with them is that they are not portable to other backends.

Reply:
I have yet to implement similar behaviour to wayland (first have to set it up, though). It may be useful if people want to make their own clipboard manager widgets that show rich previews. I suspect the targets (MIME types of the selection) will remain the same. The only thing would likely be that some statuses would become obsolete - but I would have to investigate this further. I am not familiar with wayland clipboard at all.

Reply:
I guess what we could do is try to pass the content of the selection in a generic way. What would that contain? Like you mentioned, I guess there would be a MIME type, and then the data. Currently the Wayland backend never fires these hooks but ironing this out would make it a good opportunity to do so.

Reply:
This is a somewhat difficult question to answer. Under X, the contents of a selection might change significantly depending on the required MIME type. Simplest example to try is copy something from a spreadsheet editor (such as LibreOffice) and paste it into a plain text editor. As expected, you get a text representation of the table. However, fire up GIMP and paste that - you get an image. My understanding is that it's the application providing the clipboard that handles the conversion (as it's called in the X protocol) process.

Just getting all the possible MIME types is out of the question. Applications such as firefox provide ~4 different image types alone.

The laziest solution would be to provide a hook that fires when a selection is about to be updated (would be fired at a similar stage as the current `selection_notify`) that provides the MIME types as parameters. The receiver would then specify what MIME types they require and a callback. After that, we'd somehow pass a dictionary with MIME type - value pairs back to the requesting object. This has the advantage of remaining fairly similar to the current API.

Also, now that I mention it, I should probably cap the size of the selection. As it stands, qtile just happily takes the whole selection, essentially causing the memory footprint of the operation to double. (clipboard provider + qtile) Correct me if I'm wrong, but it doesn't seem that X allows one to check the size of the selection ahead of time. As such, I'd have to just cancel the INCR transfer. Of course, the size limit could be made user configurable. 256MB default should be more than enough.

A more flexible approach, though requiring a bit more work and design, would be just passing a list of the MIME types along with a backend independent "handle" object via a new hook. Perhaps a sort of `ClipboardManager` integrated directly into the `Qtile` class? That manager object given a MIME type would then make the actual request on demand, rather than storing the clipboard globally as it is done now. The advantage is that this is the lightest and probably the most universal solution (I glanced at the wayland clipboard; should work, but I have no solid proof as of yet). The caveat is that the API would essentially totally change.